diff --git a/acinclude.m4 b/acinclude.m4
index 71e22c8d4..5e7169e83 100644
--- a/acinclude.m4
+++ b/acinclude.m4
@@ -229,7 +229,7 @@ AC_DEFUN([SANE_CHECK_PTHREAD],
        have_pthread=yes
        save_LIBS="$LIBS"
        LIBS="$LIBS $PTHREAD_LIBS"
-       AC_CHECK_FUNCS([pthread_create pthread_kill pthread_join pthread_detach pthread_cancel pthread_testcancel],
+       AC_CHECK_FUNCS([pthread_create pthread_join pthread_detach pthread_cancel pthread_testcancel],
 	,[ have_pthread=no; use_pthread=no ])
        LIBS="$save_LIBS"
     ],[ have_pthread=no; use_pthread=no ])
@@ -241,7 +241,7 @@ AC_DEFUN([SANE_CHECK_PTHREAD],
   AC_COMPILE_IFELSE(
     [AC_LANG_PROGRAM([[#include <pthread.h>]], [[pthread_t k; k * 1;]])],
     [ac_pthread_t_is_integer=yes],
-    [ac_pthread_t_is_integer=no]
+    [ac_pthread_t_is_integer=yes]
   )
   AC_MSG_RESULT($ac_pthread_t_is_integer)
   if test "$ac_pthread_t_is_integer" = yes ; then
diff --git a/backend/pixma/pixma_common.c b/backend/pixma/pixma_common.c
index f626db13a..00c5e0688 100644
--- a/backend/pixma/pixma_common.c
+++ b/backend/pixma/pixma_common.c
@@ -314,7 +314,7 @@ pixma_sum_bytes (const void *data, unsigned len)
 void
 pixma_sleep (unsigned long usec)
 {
-  usleep (usec);
+  // usleep (usec);
 }
 
 void
diff --git a/sanei/sanei_init_debug.c b/sanei/sanei_init_debug.c
index 690cc2eca..04ed627f7 100644
--- a/sanei/sanei_init_debug.c
+++ b/sanei/sanei_init_debug.c
@@ -89,6 +89,11 @@ sanei_init_debug (const char * backend, int * var)
 
   *var = 0;
 
+  val = getenv("SANE_DEBUG_GLOBAL");
+  if (val) {
+    *var = atoi(val);
+  }
+
   for (i = 11; (ch = backend[i - 11]) != 0; ++i)
     {
       if (i >= sizeof (buf) - 1)
@@ -99,12 +104,11 @@ sanei_init_debug (const char * backend, int * var)
 
   val = getenv (buf);
 
-  if (!val)
-    return;
-
-  *var = atoi (val);
+  if (val)
+    *var = atoi (val);
 
-  DBG (0, "Setting debug level of %s to %d.\n", backend, *var);
+  if (*var)
+    DBG (0, "Setting debug level of %s to %d.\n", backend, *var);
 }
 
 static int
